### DISCUSSION TITLE
Print arguments from annotations in comments

### DISCUSSION THREAD

**Discussion by phryneas** [RecordReplay/devtools#8360] [Ideas]
My main use case for this is to ship predefined sets of print statements with libraries - so a user can quickly debug what is going on inside a library without requiring Replay to add specialized Devtools for every library.

In the case of Redux, that could e.g. look something like 
```js
// replay-print({ library: 'redux', tag: 'core', severity: 'info'}, 'reducer received action %o for state %o', action, state)
state = reducer(state, action)
// replay-print({ library: 'redux', tag: 'core', severity: 'debug'}, 'state after: %o', state)
```

Replay would then extract all these print statements when parsing the source maps and display them in an extra panel where they could be toggled by the user - by library, tag and log level.
So you could just enable all "redux/core/info" logs or all "ApolloClient/cache/debug" logs with one click and without having any extra runtime code or having to find the according code passages & adding log statements yourself.

Of course, this would also enable teams to add extra log statements to their codebase without littering the normal browser console, which could be super useful.

**Comment by phryneas** [comment]
This could also be extended with further functionality, like 
```js
// replay-print-incremental-diff({ library: 'redux', tag: 'core', severity: 'debug', diffvalue: getState()}, 'action: %o', action)
```
which could then result in a panel like the Redux Devtools, showing the diff of the state from action to action.

**Comment by jasonLaster** [comment]
@phryneas very cool. This would be fairly easy for us / you / anyone to experiment with. Here's how i would tackle it

1. When a DevTools session starts, we programmatically kick off a search for `replay-print`
2. When the results come back, we parse the results and add print statements in the source files.

This would be the simple version. Of course if this is useful, we could run the analysis directly and then use the results in a redux panel or other panel.
